server/src/services/order-service.ts

```diff
@@ -0,0 +1,90 @@
+import { getCustomRepository } from 'typeorm';
+import ProductNotfoundException from '../exceptions/product-notfound-exception';
+import UserNotfoundException from '../exceptions/user-notfound-exception';
+import Order from '../models/order';
+import Product from '../models/product';
+import OrderDetailRepository from '../repositories/order-detail-repository';
+import OrderRepository from '../repositories/order-repository';
+import ProductOptionRepository from '../repositories/product-option-repository';
+import ProductRepository from '../repositories/product-repository';
+import UserRepository from '../repositories/user-repository';
+import { isNone } from '../util/type-guard';
+import OrderCreate from '../validations/order-create';
+
+const ERROR_MESSAGES = {
+  USER_NOTFOUND: '회원이 존재하지 않습니다.',
+  PRODUCT_NOTFOUND: '해당 상품이 없습니다',
+};
+
+type OrderDetailLike = {
+  discountRate: number;
+  quantity: number;
+  price: number;
+  option: string;
+  product: Product;
+};
+
+class OrderService {
+  async findByUser(userId: number): Promise<Order[]> {
+    const orders = await getCustomRepository(OrderRepository).findWithOrderDetails(userId);
+
+    return orders;
+  }
+
+  async createOrder(userId: number, orderCreate: OrderCreate): Promise<void> {
+    const { address, recipientName } = orderCreate;
+    const orderDetails: OrderDetailLike[] = [];
+
+    const user = await getCustomRepository(UserRepository).findOne(userId);
+    if (isNone(user)) {
+      throw new UserNotfoundException(ERROR_MESSAGES.USER_NOTFOUND);
+    }
+
+    let totalPrice = 0;
+    for (const orderDetail of orderCreate.orderDetails) {
+      const { optionIds, productId, quantity } = orderDetail;
+      const options = await Promise.all(
+        optionIds.map((id) => getCustomRepository(ProductOptionRepository).findOne(id))
+      );
+      const product = await getCustomRepository(ProductRepository).findOne(productId);
```

Comment:
중요한 것은 아닌데, 요 부분에서 이렇게 처리하면 조금 복잡하긴 하겠지만, options와 product는 서로 순서(?)가 상관없으니까 이렇게 불러오면 아주 조금 시간을 줄일 수 있지 않을까요 ?

```suggestion
      const optionsPromises = Promise.all(
        optionIds.map((id) => getCustomRepository(ProductOptionRepository).findOne(id))
      );

      const productPromise = getCustomRepository(ProductRepository).findOne(productId);

      const [options, product] = await Promise.all([optionsPromises, productPromise]);
```

Reply:
오 이 방법은 생각을 못했네요! 감사합니다 :)

Reply:
이차원 배열이 형성되어 잘 안되네요 ㅠㅜ

Reply:
수정했습니다 :) 꼼꼼한 리뷰 감사합니당 ㅎㅎ